src/test/scala/scala/collection/immutable/TestLazyListExtensions.scala

```diff
@@ -0,0 +1,115 @@
+
+import org.junit.Assert._
+import org.junit.Test
+
+import next._
+
+class TestLazyListExtensions {
+
+  // This method will *not* terminate for non-cyclic infinite-sized collections.
+  def assertConstantMemory[T](xs: LazyList[T]): Unit =
+    // `force` does cycle detection, so if this terminates, the collection is
+    // either finite or a cycle
+    xs.force
+
+  @Test
+  def testEmpty1(): Unit = {
+    val xs = LazyList.empty  // realized
+    val cyc = xs.cycle
+    assertTrue(cyc.isEmpty)
+    assertTrue(cyc.size == 0)
+    assertEquals(Nil, cyc.toList)
+  }
+  @Test
+  def testEmpty2(): Unit = {
+    val xs = LazyList()  // not realized
+    val cyc = xs.cycle
+    assertTrue(cyc.isEmpty)
+    assertTrue(cyc.size == 0)
+    assertEquals(Nil, cyc.toList)
+  }
+  @Test
+  def testNonEmpty(): Unit = {
+    val xs = LazyList(1, 2, 3)
+    val cyc = xs.cycle
+    assertFalse(cyc.isEmpty)
+    assertConstantMemory(cyc)
+    assertEquals(LazyList(1, 2, 3, 1, 2, 3, 1, 2), cyc.take(8))
+  }
+  @Test
+  def testToString(): Unit = {
+    assertEquals("LazyList()",
+      LazyList.empty.cycle.toString)
+    assertEquals("LazyList(<not computed>)",
+      LazyList(1, 2, 3).cycle.toString)
+    // note cycle detection here!
+    assertEquals("LazyList(1, 2, 3, <cycle>)",
+      LazyList(1, 2, 3).cycle.force.toString)
+  }
+  @Test
+  def testRepeats(): Unit = {
+    val xs = LazyList(1, 2, 3)
+    val cyc = xs.cycle
+    assertFalse(cyc.isEmpty)
+    assertEquals(LazyList(1, 2, 3, 1, 2, 3, 1, 2), cyc.take(8))
+  }
+  @Test
+  def testConstantMemory1(): Unit = {
+    val xs = LazyList(1, 2, 3)
+    val cyc = xs.cycle
+    assertTrue(cyc.tail eq cyc.tail.tail.tail.tail)
+    assertTrue(cyc.tail.tail eq cyc.drop(4).tail)
+    assertTrue(cyc.tail eq cyc.drop(3).tail)
+  }
+  @Test
+  def testConstantMemory2(): Unit = {
+    var counter = 0
+    def count(): Int = { counter += 1; counter }
+    val xs = count() #:: count() #:: count() #:: LazyList.empty
+    val cyc = xs.cycle
+    assertEquals(0, counter)
+    assertEquals(10, cyc.take(10).size)
+    assertEquals(3, counter)
+  }
+  @Test
+  def testConstantMemory3(): Unit = {
+    val xs = LazyList(1, 2, 3)
+    val cyc = xs.cycle
+    assertConstantMemory(cyc)
+    assertConstantMemory(cyc.tail)
+    assertConstantMemory(cyc.tail.tail)
+    assertConstantMemory(cyc.tail.tail.tail)
+    assertConstantMemory(cyc.tail.tail.tail.tail)
+    assertConstantMemory(cyc.drop(1))
+    assertConstantMemory(cyc.drop(10))
+  }
+  @Test
+  def testUnbounded(): Unit = {
+    val xs = LazyList.from(1)
+    val cyc = xs.cycle
+    assertEquals(LazyList(1, 2, 3), cyc.take(3))
+  }
+  @Test
+  def testSecondCallIsSafeButNotIdempotent(): Unit = {
+    val xs = LazyList(1, 2, 3)
+    // this is safe to do
+    val twice = xs.cycle.cycle
+    // and the contents are as expected
+    assertEquals(LazyList(1, 2, 3, 1, 2, 3, 1, 2), twice.take(8))
+    // but the result is not a cycle. TODO do we care?
```

Comment:
I don't think we care. If we do, I'm not sure how we'd do it. _technically_ it is currently correct (the worst kind of correct), as the first `.cycle` returns an infinite `LazyList`, so the cycle length of the second `.cycle` should be infinite

Reply:
comment updated.